### PR TITLE
feat(stream): let training.stream_vram_override replace the measured free VRAM

### LINKED ISSUE #347: Layer streaming: add an explicit override for a refused VRAM pre-flight (training.stream_vram_override)
Follow-up to #327, which stays open on the constant itself.

`LOGITS_BYTES_PER_ELEMENT = 14` is deliberately kept (see the decision comment on
#327): it is the only value under which neither published measurement grid
under-predicts, and an under-prediction on Windows/WDDM is not an exception but a
silent spill to host memory (measured: 9.27 GB allocated on a 4.29 GB card with no
exception raised).

But the asymmetry justifies the **default**, not the absence of an exit. Today a
refusal is a wall: `decide_stream_fit` returns `fits=False` and
`stream_setup._stream_budget_lines` raises. A user who knows their own stack has no
way through except lowering `max_length` or `batch_size` to a configuration they did
not want.

Measured cost of the over-prediction, DPO batch 1 / seq 768 / vocab 151936 /
rows=2 (233,373,696 logit elements, so 222.56 MiB per 1 B/elt):

| slope | predicted | vs 3166.20 MiB budget | vs 3092.46 MiB measured |
|---|---|---|---|
| 14 (shipped) | 3541.84 MiB | **refused**, −375.6 | +14.53% |
| 12.311 (this-box fit) | 3165.94 MiB | passes by 0.26 MiB | +2.38% |
| 12 (loss path alone) | 3096.72 MiB | passes by 69.5 MiB | +0.14% |

## Proposed fix

Add an explicit, opt-in override — `training.stream_vram_override` (bool) and/or a
`--stream-assume-fits` flag — that:

- prints the prediction, the measured free VRAM and the shortfall,
- names the Windows/WDDM consequence explicitly (it does not raise; it spills),
- then proceeds instead of raising.

Default stays `14`. No row of `tests/test_v07203.py::TestPeakVramReproducesTheMeasuredGrid`
is touched. The estimator keeps its "never under-predicts" property; only the
*action taken* on a predicted miss becomes the user's call.

**This is the real closure of #327 by substance:** the symptom is "configurations
that would fit are refused with no way through", and this removes the "no way
through" half without trading a documented over-prediction for an undocumented
under-prediction.

## Acceptance criteria

- [ ] Schema field with a field-validator; footgun-reject when set while
      `stream_layers` is false (mirrors the existing `stream_source`/`stream_buffers`
      gates).
- [ ] Override path prints prediction, free VRAM, shortfall, and the WDDM warning.
- [ ] Test: an over-budget config is refused without the flag and proceeds with it.
- [ ] Test: the warning text names the silent-spill behaviour (a user who reads
      "streaming is slow" would draw exactly the wrong conclusion).
- [ ] Docs + known-limitations note updated to point at the flag.

### PR DESCRIPTION
**Pull request by AmirF194**
## Root cause

`decide_stream_fit` refuses any layer-streaming step it predicts will not fit, with no way through except lowering `training.batch_size` or `data.max_length` (`trainer/stream_setup.py:_stream_budget_lines`). That refusal is deliberate for the common case: on Windows an under-prediction does not raise, it silently spills to host memory. But the free-VRAM figure it checks against comes from `torch.cuda.mem_get_info()`, a device-level driver query, and that query cannot see a per-process cap: `torch.cuda.set_per_process_memory_fraction`, a MIG slice, or a card another process is also using. On that hardware the pre-flight sees the whole device's free VRAM regardless of what the process can actually use, so there was no way to make the pre-flight simulate the real constraint either.

## Fix

`training.stream_vram_override` (bytes) replaces the measured free-VRAM figure the pre-flight checks against, in either direction:

- raised past a documented over-prediction, to let a known-safe config through;
- lowered below the real free VRAM, to enforce a cap the driver itself cannot report (a capped Colab/Kaggle T4, a MIG slice).

Rejected at config load when set while `stream_layers` is false, mirroring the existing `stream_source`/`stream_buffers` footgun gate.

**One note on shape, since the issue changed under me while it was open.** The original acceptance criteria describe a `bool`/`force` flag that lets an over-budget run proceed. The follow-up comment on this issue narrows and extends that: a boolean cannot express "assume only 4 GB is free" for the capped-notebook case, so it asks for a numeric value that *replaces* the measured figure instead. I built the follow-up comment's shape, since it is the one the notebook use case actually needs and it still satisfies the original ask (an over-prediction can be raised past). One consequence: the override does not unconditionally let a run "proceed" the way the original bool would have. It changes the yardstick `decide_stream_fit` measures against, and that comparison still refuses if the predicted peak exceeds it, which is what makes "override set below real free VRAM refuses a fitting config" a real test rather than a tautology.

## Verification

- New tests fail on `main` (`ImportError`, `resolve_available_vram_bytes` does not exist there) and pass on this branch: `resolve_available_vram_bytes` unit tests (no override uses the measured figure; override replaces a larger or smaller measured figure either way), the acceptance test the follow-up comment describes verbatim (override set below real free VRAM turns an otherwise-fitting config into a refusal), and schema tests (default `None`, accepts an explicit byte count, rejects `bool`/negative, rejects being set while `stream_layers` is false).
- `pytest tests/test_v07200.py tests/test_v07201.py tests/test_v07202.py tests/test_v07203.py tests/test_v07204.py tests/test_v07300.py` (the layer-streaming and preference-loss suites, 476 tests): all pass, 21 skipped (CUDA-gated, expected offline). `ruff check src/soup_cli/ tests/` clean.
- Not verified: the trainer-level wiring in `_stream_budget_lines` (the `on_cuda` branch that actually calls `resolve_available_vram_bytes`) has no GPU available in this sandbox, so its two CUDA-gated integration tests are skipped, the same as every other test in this file that drives that branch. Only Python 3.12 of this repo's 3.10/3.11/3.12 CI matrix was run.

Fixes #347

### CODE REVIEW

**Review by MakazhanAlpamys — CHANGES_REQUESTED** [review]
Six hours, and the judgement call in it is right. Ran everything rather than reading it.

## The shape decision

You took the follow-up comment's shape over the issue's original acceptance criteria and said so explicitly, with the reason. That is the correct call and the correct way to make it — a `force: true` would have closed #347 by its own text and left the notebook impossible, which is how an issue gets closed while the thing it was blocking stays blocked.

Your note that the override "does not unconditionally let a run proceed" is the part I would have written the review around if you had not. It changes the yardstick rather than skipping the check, and that is exactly what keeps *"override below real free VRAM refuses a fitting config"* from being a tautology. Nothing else in this diff would have told me you understood that distinction; the sentence does.

Two details I did not ask for and would not have thought to:

- **The panel prints both numbers** — `free VRAM 4.00 GB (training.stream_vram_override; driver reports 15.36 GB)`. Without it, the one line a user reads before a long run silently means something different than it did yesterday.
- **`ge=0` with `0` allowed.** Right: "assume nothing is free" is a real request, and it is the value someone reaches for to check the flag is wired at all.

## Verified

```
branch, tests/test_v07200.py + tests/test_v07203.py   284 passed, 2 skipped
same tests against main's source                        6 failed, 113 passed
ruff check src/soup_cli/ tests/                         clean
```

`resolve_available_vram_bytes` lives in `layer_stream.py` (planner, torch-free) rather than in the trainer, which keeps the light CLI path light and the function unit-testable without CUDA. Correct home.

## Blocking, and it is one line

**The suite passes with the `is None` check replaced by truthiness.** I mutated it:

```
$ # return override_bytes or measured_bytes
284 passed, 2 skipped
```

Nothing fails. `test_zero_is_accepted` proves the value survives config load; no test proves it survives `resolve_available_vram_bytes`. Under the mutation `stream_vram_override: 0` silently falls back to the driver figure — so the user asks for "assume nothing is free, refuse everything", the run proceeds, and the pre-flight reports the opposite of what was requested.

That value is not hypothetical. `0` is precisely what someone sets to check the flag is wired at all, expecting a refusal — the one input where a silent no-op reads as confirmation. You allowed `0` deliberately at the schema; the resolver should be pinned to honour it:

```python
assert resolve_available_vram_bytes(measured_bytes=15_360_000_000, override_bytes=0) == 0
```

My mutation then fails, and the `is None` you already wrote is defended rather than incidental.

## Non-blocking: the flag is not discoverable where it is needed

`docs/performance-and-quantization.md` gains one line, in the footgun list. But the two YAML config blocks a reader actually copies from (`:236-237` and `:374-375`) list `stream_source` and `stream_buffers` with inline comments and do not mention this field — so someone reading the streaming config reference will not learn it exists.

More importantly, there is no entry against the message where a user meets this problem. Someone hits `predicted peak … exceeds free VRAM` and their options today are "lower `batch_size`" or "lower `max_length`" — the two the issue says are not what they wanted. The troubleshooting list at `:398` is where the third option belongs, with the warning attached: raising it past a real limit on Linux is an OOM, and on Windows it is a silent spill to host memory rather than an exception.

One line in each config block and one troubleshooting entry. Happy to take it myself after merge if you would rather keep the PR at its current scope — say which.

---

The test is the only thing I need. CI is green on lint, type-check and recipes with the matrix still running; I will merge on your next push.

**Comment by AmirF194** [comment]
Added the exact assertion at 50eba84: `resolve_available_vram_bytes(measured_bytes=15_360_000_000, override_bytes=0) == 0`. Reproduced your mutation locally first (`override_bytes or measured_bytes`, `284 passed, 2 skipped`, the assertion's own value comes back `15360000000` instead of `0`), confirmed the new test fails on it, then confirmed it passes against the real `is None` check.

Also rebased onto main (the bf16 fix landed in the same window) -- one conflict in CHANGELOG.md, both entries kept (Added before Fixed). Full suite still green: 477 passed, 21 skipped; ruff clean.

Left the docs placement as a follow-up per your offer, thanks for taking it.

**Review by MakazhanAlpamys — APPROVED** [review]
All addressed, and the mutation now dies where it should.

```
$ # return override_bytes or measured_bytes
FAILED ...::TestResolveAvailableVramBytes::test_override_zero_is_honoured_not_treated_as_absent
1 failed, 119 passed
```

One test, the right one, and nothing else moved — the `is None` you already had is now defended rather than incidental. That you reproduced the mutation first and checked the assertion's own value came back `15360000000` instead of `0` is the part that makes it a real check: you confirmed the test fails for the reason you think it does, not just that it fails.

CI green, 12/12. Merging.

Docs placement is mine — the two YAML config blocks in `docs/performance-and-quantization.md` and a troubleshooting entry against the refusal message, since that is where a user actually meets this.

**Comment by MakazhanAlpamys** [comment]
Merged as [`d31c1e5`](https://github.com/MakazhanAlpamys/Soup/commit/d31c1e5), 12/12 green. #347 closed.

Docs follow-up landed in [`5752249`](https://github.com/MakazhanAlpamys/Soup/commit/5752249), as offered — the field is now in the streaming config block plus two troubleshooting entries: the over-prediction case (with the warning that raising it past a real limit is an OOM on Linux and a *silent spill* on Windows), and the capped/shared-GPU case, which is the one `mem_get_info` structurally cannot see. Your second entry is in [`CONTRIBUTORS.md`](https://github.com/MakazhanAlpamys/Soup/blob/main/CONTRIBUTORS.md).

Two PRs in three days, both of which raised the bar on the ask rather than meeting it: #384 chose (B) explicitly and wrote down why, this one preferred the follow-up comment's shape over the issue's own acceptance criteria and argued the case. The second is the harder call — an issue's text is the safe thing to build against, and building the thing it *should* have said, out loud, is how a stale acceptance criterion gets fixed instead of satisfied.

**#347 was blocking something outside the tracker**, which is now unblocked by this: a free-tier Colab notebook where a reader caps a T4 to ~4 GB and watches a streamed 8B run agree with a resident reference. That was impossible before this PR for exactly the reason in your root-cause paragraph.

#348 next, whenever suits. The guardrail from the issue comment still stands — it is not a route around #327's decision, the constant stays at 14; what is dead is the opt-in upward-only calibration, and the hole it was written to catch is still open.

**Comment by AmirF194** [comment]
Thanks, and appreciate you taking the docs follow-up too.
